clover.rb

```diff
@@ -3,38 +3,588 @@
+    flash["error"] = "An invalid security token submitted with this request, please try again"
+    return redirect_back_with_inputs
+  end
+
+  plugin :content_security_policy do |csp|
+    csp.default_src :none
+    csp.style_src :self, "https://cdn.jsdelivr.net/npm/flatpickr@4.6.13/dist/flatpickr.min.css"
+    csp.img_src :self, "data: image/svg+xml"
+    csp.form_action :self, "https://checkout.stripe.com"
+    csp.script_src :self, "https://cdn.jsdelivr.net/npm/jquery@3.7.0/dist/jquery.min.js", "https://cdn.jsdelivr.net/npm/dompurify@3.0.5/dist/purify.min.js", "https://cdn.jsdelivr.net/npm/flatpickr@4.6.13/dist/flatpickr.min.js"
+    csp.connect_src :self
+    csp.base_uri :none
+    csp.frame_ancestors :none
+  end
+
+  # XXX: Temporary unless PR 2084 is merged
+  # :nocov:
+  if Config.production?
+    def api?
+      return @is_api if defined?(@is_api)
+      @is_api = env["HTTP_HOST"]&.start_with?("api.")
+    end
+  else
+    def api?
+      return @is_api if defined?(@is_api)
+      @is_api = env["HTTP_HOST"]&.start_with?("api.") || env["PATH_INFO"].start_with?("/api")
+    end
+  end
+  # :nocov:
+
+  def runtime?
+    return @is_runtime if defined?(@is_runtime)
+    @is_runtime = env["PATH_INFO"].start_with?("/runtime")
+  end
+
+  def web?
+    return @is_web if defined?(@is_web)
+    @is_web = !api? && !runtime?
+  end
+
+  def get_jwt_payload(request)
+    return unless (v = request.env["HTTP_AUTHORIZATION"])
+    jwt_token = v.sub(%r{\ABearer:?\s+}, "")
+    begin
+      JWT.decode(jwt_token, Config.clover_runtime_token_secret, true, {algorithm: "HS256"})[0]
+    rescue JWT::DecodeError
+    end
+  end
+
+  def csrf_tag(*)
+    render("components/form/hidden", locals: {name: csrf_field, value: csrf_token(*)})
+  end
+
+  def redirect_back_with_inputs
+    flash["old"] = request.params
+    request.redirect env["HTTP_REFERER"] || "/"
+  end
+
+  def has_project_permission(actions)
+    @project_permissions.intersection(Authorization.expand_actions(actions)).any?
+  end
+
+  plugin :not_found do
+    @error = {
+      code: 404,
+      type: "ResourceNotFound",
+      message: "Sorry, we couldn’t find the resource you’re looking for."
+    }
+
+    if api?
+      {error: @error}.to_json
+    else
+      view "/error"
+    end
+  end
+
+  if Config.development?
+    # :nocov:
+    plugin :exception_page
+
+    class RodaRequest
+      def assets
+        exception_page_assets
+        super
+      end
+    end
+    # :nocov:
+  end
+
+  plugin :error_handler do |e|
+    # raise e
+    error = parse_error(e)
+
+    if api? || runtime?
+      unless runtime? && error[:code] == 204
+        {error: error}.to_json
+      end
+    else
+      @error = error
+
+      case e
+      when Sequel::ValidationFailed
+        flash["error"] = @error[:message]
+        return redirect_back_with_inputs
+      when Validation::ValidationFailed
+        flash["errors"] = (flash["errors"] || {}).merge(@error[:details])
```

Comment:
Actually we do that. `redirect_back_with_inputs` puts old values to flash. Our form components check old value from flash

Reply:
Ah.  The issue with that is that large form inputs will break things, since flash is stored in the session cookie, so a large form input would exceed the 4K cookie limit and and break the session cookie.